connectors/services/attachments.py

```diff
@@ -0,0 +1,250 @@
+        target = os.path.join(self.drop_dir, key)
+        async with aiofiles.open(target, "wb") as f:
+            async for chunk in gen:
+                await f.write(chunk)
+
+        logger.info(f"Dropped {key}")
+
+        # the file is now on disk, let's create the desc
+        desc = {
+            # XXX maybe its overkill for v1 since a node only connects to a
+            # single ES
+            # XXX add suport for API key
+            "filename": key,
+            "index": index,
+            # XXX how do we know this value when the initial sync of the doc is
+            # not done yet. we might need to use the source id
+            # and at sync time, query for the corresponding ES doc id
+            # and send it only once it's there..
+            "doc_id": doc_id,
+            "name": filename,
+        }
+
+        desc_file = os.path.join(self.drop_dir, os.path.splitext(key)[0] + ".json")
+        with open(desc_file, "w") as f:
+            f.write(json.dumps(desc))
+
+        logger.info(f"Dropped {desc_file}")
+
+
+class AttachmentService(BaseService):
+    """Watches a directory for files and sends them by chunks
+
+    Uses `Transfer-Encoding: chunked`
+    """
+
+    def __init__(self, args):
+        super().__init__(args)
+        self.elastic_config = self.config["elasticsearch"]
+        self._config = self.config.get("attachments", {})
+        self.max_concurrency = self._config.get("max_concurrency", 5)
+        self.directory = self._prepare_dir("drop")
+        self.idle_time = self._config.get("idling", 30)
+        self.log_directory = self._prepare_dir("logs")
+        self.results_directory = self._prepare_dir("results")
+        self.running = False
+        self._sleeps = CancellableSleeps()
+
+    def _to_b64(self, filename):
+        """Calls the system base64 utility to create a base64-encoded file
+
+        The base64 utility is a stream encoder, the file will not be fully
+        loaded into memory.
+
+        This is a blocking method.
+        """
+        cmd = f"{BASE64} {filename} > {filename}.b64"
+        logger.debug(f"[{os.path.basename(filename)}] Running {cmd}")
+        os.system(cmd)
+        return f"{filename}.b64"
+
+    async def _file_to_pipeline(self, filename, chunk_size=1024 * 1024):
+        """Convert a file into a streamable Elasticsearch request.
+
+        - Calls `_to_b64` in a separate thread so we don't block
+        - Reads the base64 file by chunks an provide an async generator
+        """
+        b64_file = await get_event_loop().run_in_executor(None, self._to_b64, filename)
+
+        # XXX adding .txt to all known text files
+        if os.path.splitext(filename)[-1] in SUPPORTED_TEXT_FILETYPE:
+            filename += ".txt"
+
+        try:
+            async with aiofiles.open(b64_file, "rb") as f:
+                yield b'{"filename":"' + filename.encode("utf8") + b'","data":"'
+                chunk = (await f.read(chunk_size)).strip()
```

Comment:
Not sure if this service supports Python 3.8+ only, but if so this is the perfect opportunity for a walrus operator:

```python
while chunk := (await f.read(chunk_size)).strip()):
    yield chunk
```